domain/externalcontroller/modelmigration/import.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelmigration
+
+import (
+	"context"
+
+	"github.com/juju/description/v4"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/modelmigration"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/externalcontroller"
+	"github.com/juju/juju/domain/externalcontroller/service"
+	"github.com/juju/juju/domain/externalcontroller/state"
+)
+
+// Coordinator is the interface that is used to add operations to a migration.
+type Coordinator interface {
+	// Add adds the given operation to the migration.
+	Add(modelmigration.Operation)
+}
+
+// RegisterImport registers the import operations with the given coordinator.
+func RegisterImport(coordinator Coordinator) {
+	coordinator.Add(&importOperation{})
+}
+
+type ImportService interface {
+	ImportExternalControllers(context.Context, []externalcontroller.MigrationControllerInfo) error
+}
+
+type importOperation struct {
+	modelmigration.BaseOperation
+
+	service ImportService
+}
+
+func (i importOperation) Setup(scope modelmigration.Scope) error {
+	// We must not use a watcher during migration, so it's safe to pass a
+	// nil watcher factory.
+	i.service = service.NewService(
+		state.NewState(domain.ConstFactory(scope.ControllerDB())), nil)
+	return nil
+}
+
+func (i importOperation) Execute(ctx context.Context, model description.Model) error {
+	externalControllers := model.ExternalControllers()
+	if len(externalControllers) == 0 {
+		return nil
+	}
+
+	var controllers []externalcontroller.MigrationControllerInfo
+	for _, entity := range externalControllers {
+		controllers = append(controllers, externalcontroller.MigrationControllerInfo{
+			ControllerTag: entity.ID(),
+			Alias:         entity.Alias(),
+			CACert:        entity.CACert(),
+			Addrs:         entity.Addrs(),
+			ModelUUIDs:    entity.Models(),
+		})
+	}
+
+	err := i.service.ImportExternalControllers(ctx, controllers)
```

Comment:
As noted later on, I think we should just call `UpdateExternalController` per external controllers. Although that's running multiple transactions, we don't actually care.

Reply:
So, what I decided to do is to keep this separate function to be 100% consistent to what we had in the legacy state (i.e. having all imports under one single transaction). I prefer to be conservative in this case and it makes sense to have everything under one tx.
I have refactored a bit to have the core of the db operations in a separate method that's called both by `UpdateExternalController` and `ImportExternalControllers`.
If you still prefer, I can rollback and user multiple TXs by calling multiple times `UpdateExternalController()`.